modules/animation/src/datastructures/animation.cpp

```diff
@@ -71,13 +73,95 @@ void Animation::add(std::unique_ptr<Track> track) {
     priorityTracks_.push_back(tracks_.back().get());
     doPrioritySort();
     tracks_.back()->addObserver(this);
+    if (auto propertyTrack = dynamic_cast<BasePropertyTrack*>(tracks_.back().get())) {
+        propertyTrack->getProperty()->getOwner()->addObserver(this);
+    }
     notifyTrackAdded(tracks_.back().get());
 }
 
+Keyframe* Animation::addKeyframe(Property* property, Seconds time) {
+    auto it = findTrack(property);
+    try {
+        if (it != end()) {
+            // Note: interpolation will only be used if a new sequence is created.
+            return dynamic_cast<BasePropertyTrack*>(&(*it))->addKeyFrameUsingPropertyValue(time);
+        } else if (auto basePropertyTrack = add(property)) {
+            return basePropertyTrack->addKeyFrameUsingPropertyValue(time);
+        } else {
+            LogWarn("No matching Track found for property \"" << property->getIdentifier() << "\"");
+        }
+    } catch (const Exception& ex) {
+        LogError(ex.getMessage());
+    }
+    return nullptr;
+}
+
+KeyframeSequence* Animation::addKeyframeSequence(Property* property, Seconds time) {
+    auto it = findTrack(property);
+    std::string interpolationErrMsg;
+    try {
+        if (it != end()) {
+            return dynamic_cast<BasePropertyTrack*>(&(*it))->addSequenceUsingPropertyValue(time);
+        } else if (auto basePropertyTrack = add(property)) {
+            basePropertyTrack->addKeyFrameUsingPropertyValue(time);
+            return &basePropertyTrack->toTrack()->getFirst();
+        } else {
+            LogWarn("No matching Track found for property \"" << property->getIdentifier() << "\"");
+        }
+    } catch (const Exception& ex) {
+        LogError(ex.getMessage());
+    }
+    return nullptr;
+}
+
+Animation::iterator Animation::findTrack(Property* property) {
+    return std::find_if(begin(), end(), [property](auto& track) {
+        if (auto pTrack = dynamic_cast<BasePropertyTrack*>(&track)) {
+            return pTrack->getProperty() == property;
+        } else {
+            return false;
+        }
+    });
+}
+
+BasePropertyTrack* Animation::add(Property* property) {
+    auto it = findTrack(property);
+    if (it != end()) {
+        return dynamic_cast<BasePropertyTrack*>(&(*it));
+    } else {
+        if (auto track = getManager()->getTrackFactory().create(property)) {
```

Comment:
should we not start observing the owner? or where does that happen?

Reply:
We do when the track has been added, `add(std::unique_ptr<Track>)`